src/main/java/com/ontotext/kafka/processor/SinkRecordsProcessor.java

```diff
@@ -0,0 +1,263 @@
+package com.ontotext.kafka.processor;
+
+import com.ontotext.kafka.GraphDBSinkConfig;
+import com.ontotext.kafka.error.LogErrorHandler;
+import com.ontotext.kafka.processor.record.handler.RecordHandler;
+import com.ontotext.kafka.processor.retry.GraphDBRetryWithToleranceOperator;
+import com.ontotext.kafka.rdf.repository.RepositoryManager;
+import com.ontotext.kafka.util.ValueUtil;
+import org.apache.kafka.connect.errors.ConnectException;
+import org.apache.kafka.connect.errors.RetriableException;
+import org.apache.kafka.connect.runtime.errors.Stage;
+import org.apache.kafka.connect.sink.SinkRecord;
+import org.eclipse.rdf4j.repository.RepositoryConnection;
+import org.eclipse.rdf4j.repository.RepositoryException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.slf4j.MDC;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Queue;
+import java.util.UUID;
+import java.util.concurrent.ConcurrentLinkedQueue;
+import java.util.concurrent.LinkedBlockingQueue;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+/**
+ * A processor which batches sink records and flushes the updates to a given
+ * GraphDB {@link org.eclipse.rdf4j.repository.http.HTTPRepository}.
+ * <p>
+ * Batches that do not meet the {@link com.ontotext.kafka.GraphDBSinkConfig#BATCH_SIZE} are flushed
+ * after passing {@link com.ontotext.kafka.GraphDBSinkConfig#RECORD_POLL_TIMEOUT} threshold.
+ *
+ * @author Tomas Kovachev tomas.kovachev@ontotext.com
+ */
+public final class SinkRecordsProcessor implements Runnable {
+
+	private static final Logger LOG = LoggerFactory.getLogger(SinkRecordsProcessor.class);
+
+	private final UUID id = UUID.randomUUID();
+	private final LinkedBlockingQueue<Collection<SinkRecord>> sinkRecords;
+	private final Queue<SinkRecord> recordsBatch;
+	private final RepositoryManager repositoryManager;
+	private final int batchSize;
+	private final long timeoutCommitMs;
+	private final LogErrorHandler errorHandler;
+	private final RecordHandler recordHandler;
+	private final GraphDBSinkConfig.TransactionType transactionType;
+	private final GraphDBSinkConfig config;
+	private final AtomicBoolean backOff = new AtomicBoolean(false);
+
+	public SinkRecordsProcessor(GraphDBSinkConfig config) {
+		this(config, new LinkedBlockingQueue<>(), new RepositoryManager(config));
+	}
+
+	public SinkRecordsProcessor(GraphDBSinkConfig config, LinkedBlockingQueue<Collection<SinkRecord>> sinkRecords, RepositoryManager repositoryManager) {
+		this.config = config;
+		this.sinkRecords = sinkRecords;
+		this.recordsBatch = new ConcurrentLinkedQueue<>();
```

Comment:
Agreed. This can be a LinkedList. 
Just a note - the task is not single-threaded. When you create a connector you can specify the number of tasks for that connector. These are threaded instances of the same SinkTask. 
The processor is intended to a be single instance for all task threads of the same connector.